docs/research/blake-256-hash-function.md

```diff
@@ -1,13 +1,302 @@
+
+BLAKE's uses HAIFA, which maintains the valuable properties of the M-D construction and adds to the security and scalability of the transformation. HAIFA is essentially M-D construction but with a mandatory counter (number of bits hashed so far) and an optional salt (random data that used as an additional input).
+
+![HAIFA Construction](/img/HAIFA (BLAKE).svg)
+
+**<em>Figure 2: HAIFA Construction</em>**
+
+This iteration mode solves many of the internal collision problems with M-D construction and provides resistance to herding, long message second pre-image, and length extension attacks as the additional salt and counter simulate distinct compression functions for each data block processed [^4] [^12]. HAIFA is both prefix- and suffix-free and, as a result, is collision-resistant as well as indifferentiable from a random oracle, assuming the underlying compression function is ideal [^1] [^13] [^14]. 
+
+#### **The local wide-pipe compression function**
+
+
+![Local wide-pipe](/img/Wide-pipe.svg)
+
+**<em>Figure 3: Local wide-pipe construction of BLAKE's compression function, inherited from LAKE hash function [^6]</em>**
+
+BLAKE uses the LAKE local wide-pipe structure for its strong security guarantees against collision attacks [^1] [^6]. The compression function takes as input four values: a chaining value, a message block, an optional salt and a mandatory counter. In the case of BLAKE-256: 
+
+| chaining value| message block| Salt  | counter
+|---------------|--------------|-------|--------|
+|     256-bit   |   512-bit    |128-bit|64 -bit |
+
+The structure then works in three steps:
+
+1) Initialization of a large internal state represented by 16 words in a 4x4 word matrix.
+
+2) Iteration of 14 rounds, each round that calls the core function (**G** as explained in the next section) 8 times to transform this state.
+
+3) Finalization after a series of rounds, a new sequence of values taken from the matrix using sequence of initial values and salt. 
+
+A more detailed description of the compression function can be found in [SHA-3 proposal BLAKE](https://decred.org/research/aumasson2010.pdf).
+
+#### **ChaCha inspired core function**
+
+The core function of BLAKE is a variation of the stream cipher ChaCha by Bernstein, one of the fastest stream ciphers. ChaCha inspired the creators of BLAKE due to its simplicity and security [^1]. The core function called the **G** function is repeated 112 times in BLAKE-256 over 14 rounds with each round consisting of eight round-dependent transformations **G<sub>0</sub>**, · · · , **G<sub>7</sub>**. The **G** function of BLAKE-256 for index **i**, where **σ<sub>r</sub>** is a fixed permutation used in
```

Comment:
I'd suggest wrapping formula and formula components in `$` to make use of MathJax. 
For example:

```
The $G_i$ updates 4...
```

More examples on the `Elliptic Curve Signature Algorithms` page